src/rules_pygen/rules_generator.py

```diff
@@ -0,0 +1,514 @@
+    \.whl|\.dist-info)$""",
+    re.VERBOSE,
+)
+
+ARCHIVE_TMPL = """
+    if "{archive_name}" not in existing_rules:
+        http_archive(
+            name = "{archive_name}",
+            urls = ["{url}"],
+            sha256 = "{sha256}",
+            build_file_content = _BUILD_FILE_CONTENT,
+            type = "zip",
+        )
+"""
+
+BLACKLIST = {
+    'setuptools',
+    'typing',  # causes issues in python versions > 3.4
+}
+
+SUPPORTED_PYTHON3_VERSIONS = ['py3', 'py2.py3', 'py34', 'py35', 'py36', 'py37', 'py38']
+CPYTHON_VERSIONS = ('cp34', 'cp35', 'cp36', 'cp37', 'cp38')
+
+
+class PyBazelRuleGeneratorException(Exception):
+    pass
+
+
+def _space(x: int):
+    return ' ' * x
+
+
+def _download(url: str, dest: str) -> None:
+    logger.info('Downloading %s', url)
+    urllib.request.urlretrieve(url, dest)
+
+
+def _calc_sha256sum(filepath: str) -> str:
+    with open(filepath, 'rb') as fd:
+        digest = hashlib.sha256()
+        while True:
+            buf = fd.read(4096)
+            if not buf:
+                break
+            digest.update(buf)
+        return digest.hexdigest()
+
+
+def _check_compatibility(filename: str, desired_pyver: str) -> bool:
+    pyver, abi = '', ''
+    match  = WHEEL_FILE_RE.search(filename)
+
+    if match:
+        pyver = match.group('pyver')
+        abi = match.group('abi')
+    else:
+        raise PyBazelRuleGeneratorException("Could not get Python version information from wheel file: {}".format(filename))
+
+    supported_abis = ['abi3', 'cp{}m'.format(desired_pyver), 'cp{}mu'.format(desired_pyver)]
+    always_supported_pyvers = ['cp{}'.format(desired_pyver), 'py{}'.format(desired_pyver)]
+
+    if abi == 'none' and pyver in SUPPORTED_PYTHON3_VERSIONS:
+        return True
+    elif abi in supported_abis and pyver in CPYTHON_VERSIONS:
+        return True
+    elif pyver in always_supported_pyvers:
+        return True
+    else:
+        logger.info('Skipping, version: %s, abi: %s', pyver, abi)
+    return False
+
+
+class WheelInfo:
+    """Struct for information on a wheel."""
+
+    def __init__(self, filepath, url, name, version):
+        self.filepath = filepath
+        self.url = url
+        self.name = name.lower()
+        self.version = version
+
+        self.sha256sum = _calc_sha256sum(filepath)
+        self.filename = os.path.basename(filepath)
+
+    def __repr__(self):
+        return '<{} ({})>'.format(self.filename, self.platform)
+
+    @property
+    def platform(self) -> str:
+        if 'linux' in self.filename:
+            return 'linux'
+        elif 'macos' in self.filename:
+            return 'macos'
+        return 'purelib'
+
+    def __eq__(self, other):
+        return isinstance(other, WheelInfo) and (self.filename == other.filename)
+
+    def __hash__(self):
+        return hash(self.filename)
+
+    @property
+    def archive_name(self) -> str:
+        """Name of the archive
+
+        Example: 'pypi__futures__3_1_1'
```

Comment:
I think it should be `pypi__futures_3_1_1` (i.e. with one underscore after `futures`)